### ISSUE TITLE
Major version bump

### ISSUE THREAD

**Issue by dguest** [lwtnn/lwtnn#153]
MR #149 would switch the json parser from boost to nlohmann. Since we only have two dependencies this is a pretty big change, definitely enough to warrant a major version bump. But while we're there, I might clean some things up.

Here's what I had in mind:
- Replace boost dependency with nlohmann
- Delete non-templated versions of the classes in `generic/`
- Move templated classes out of `generic/` and into the top level, remove the `generic::` namespace
- Change the namespace of the project, from `lwt::` to `lw3`, or `lwt::v3::` or something, to make it possible to run both versions side-by-side
- Only support C++17 and up (drop support for C++11, C++14)

In short I would introduce a lot of breaking changes while making it possible for people to keep using the old version, even if they share a codebase with people using the new version. I appreciate any feedback on the best way to do this.

**Comment by matthewfeickert** [comment]
> * Change the namespace of the project, from `lwt::` to `lw3`, or `lwt::v3::` or something, to make it possible to run both versions side-by-side

I like basically everything about this, except for the fact that you're making versioned APIs just so things in Athena won't break, which now will require:
- All users who want to track the project development to change their API usage regardless of if they need to (this is a pretty small ask though)
- For `ltwnn` as a project to have to continue to support deprecated APIs as long as they are used in Athena &mdash; which is basically forever. This seems like it will now slow development because you'll need to forever test and support both APIs even though the whole point of the new release is to rid yourself of the old. Am I missing something basic here?

I admittedly don't have as much experience dealing with these integration issues with large monolith projects, so I think that @lukasheinrich and @jpivarski can probably give better advice.

**Comment by dguest** [comment]
Yeah I'd be curious about why @jpivarski decided to put the major version into the API for e.g. uproot, and whether that was a good idea in hindsight. 

One thing to keep in mind here is that lwtnn is an extremely low-maintenance project, both for me as the developer and for users. So yes, versioning the API is _just_ so things in Athena won't break, but not breaking the API one of the central goals of the project.

**Comment by jpivarski** [comment]
Putting the major version into the package name, i.e. "pip install uproot3" and "pip install uproot4", was a mistake. What I hadn't anticipated was that downstream dependencies would pin to specific versions (in projects that update over time, not just one-time tutorials), and then when "pip install uproot" switched from meaning "uproot3" to meaning "uproot4", there were version + name combinations that didn't work. It didn't matter that it was advertised months in advance, and the diagrams I made of combinations that would work and combinations that would not work didn't help, either. It's just a bad idea: the package namespace must never change.

What probably would have been fine would be to put the old and new APIs within a single package, such that old API users have to

```python
import uproot.v3 as uproot
```

from a submodule instead of a separate package. I didn't get the impression that anyone had any problems with the

```python
import uproot3 as uproot
```

one-line change to keep old code alive. Anyone who didn't know that the API change was coming found out about it through one of a small set of error messages that I recognized and could immediately recommend the one-line change; there were no complaints thereafter. So "put old codebase in a submodule" probably would have been fine, whereas "swap package names in PyPI" was bad.

Starting a new GitHub repo has not caused any problems at all. This can be considered a "development detail," which has little to no user-level consequences. On the other hand, rebooting git history cleaned up a lot of mess and sped up "git clone" times.

In a C++ project, the packaging problem may be moot because I don't think there's a global namespace of packages that is universally used among C++ projects. The idea of putting ltwnn in a versioned namespace like `lwt::v3::` doesn't sound bad to me, as I've said putting uproot3 in a submodule probably would have been fine. Users who update and get compiler errors would ask, "What happened?" and you can tell them about the one-line change they have to make to rename namespaces. It might be in a lot of files, though (e.g. at the top of every .cpp file).

**Comment by matthewfeickert** [comment]
> One thing to keep in mind here is that lwtnn is an extremely low-maintenance project, both for me as the developer and for users. So yes, versioning the API is _just_ so things in Athena won't break, but not breaking the API one of the central goals of the project.

> The idea of putting ltwnn in a versioned namespace like `lwt::v3::` doesn't sound bad to me, as I've said putting uproot3 in a submodule probably would have been fine. Users who update and get compiler errors would ask, "What happened?" and you can tell them about the one-line change they have to make to rename namespaces.

If this is the best way to update and break APIs inside of a monolith like Athena then I guess moving forward with this should work. 

My concern is that this (having people rely on versioning in the API and not rely on versioning a release) ends up keeping the current `v2` APIs around in the main codebase forever (and not just on a `v2` release branch) because some Athena user might at some point want to use it even though it is fully deprecated. This concern is less practical then the concerns of getting new APIs into Athena to use _now_, but the idea of having to keep around `v2`, `v3`, `v4`,... APIs forever in the code base seems like a bit of an anti-pattern for a project under version control.

Am I missing something here? It might be that I don't (and I know I don't) fully understand the Athena Externals system for integration of things like `ltwnn`.
